golem-worker-executor-base/src/services/worker_fork.rs

```diff
@@ -0,0 +1,178 @@
+
+        let owned_source_worker_id = OwnedWorkerId::new(account_id, source_worker_id);
+
+        let metadata = self
+            .all
+            .worker_service()
+            .get(&owned_source_worker_id)
+            .await
+            .ok_or(GolemError::worker_not_found(source_worker_id.clone()))?;
+
+        Ok((owned_source_worker_id, owned_target_worker_id, metadata))
+    }
+}
+
+#[async_trait]
+impl<Ctx: WorkerCtx, Svcs: HasAll<Ctx> + Send + Sync + 'static> WorkerFork
+    for DefaultWorkerFork<Ctx, Svcs>
+{
+    async fn fork(
+        &self,
+        source_worker_id: &OwnedWorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(), GolemError> {
+        record_worker_call("fork");
+
+        let (owned_source_worker_id, owned_target_worker_id, source_worker_metadata) = self
+            .validate_worker_forking(
+                &source_worker_id.account_id,
+                &source_worker_id.worker_id,
+                target_worker_id,
+                oplog_index_cut_off,
+            )
+            .await?;
+
+        let target_worker_id = owned_target_worker_id.worker_id.clone();
+        let account_id = owned_target_worker_id.account_id.clone();
+
+        // Not sure if we should copy the metadata or not, or stick on to just default
+        let target_worker_metadata = WorkerMetadata {
+            worker_id: target_worker_id.clone(),
+            account_id,
+            env: source_worker_metadata.env.clone(),
+            args: source_worker_metadata.args.clone(),
+            created_at: Timestamp::now_utc(),
+            parent: source_worker_metadata.parent.clone(),
+            last_known_status: WorkerStatusRecord::default(),
+        };
+
+        let source_worker_instance = Worker::get_or_create_suspended(
+            &self.all,
+            &owned_source_worker_id,
+            Some(source_worker_metadata.args.clone()),
+            Some(source_worker_metadata.env.clone()),
+            None,
+            None,
+        )
+        .await?;
+
+        let source_oplog = source_worker_instance.oplog();
+
+        let initial_oplog_entry = source_oplog.read(OplogIndex::INITIAL).await;
+
+        // Update the oplog initial entry with the new worker
+        let target_initial_oplog_entry = initial_oplog_entry
+            .update_worker_id(&target_worker_id)
+            .ok_or(GolemError::unknown(
+                "Failed to update worker id in oplog entry",
+            ))?;
+
+        let new_oplog = self
+            .all
+            .oplog_service()
+            .create(
+                &owned_target_worker_id,
+                target_initial_oplog_entry,
+                target_worker_metadata,
+                Arc::new(RwLock::new(ExecutionStatus::Suspended {
+                    last_known_status: WorkerStatusRecord::default(), // default is idle, TODO: check if we need to update this or derive from the source last known status
```

Comment:
I think this does not matter because we are going to drop this value here and the "resume" will recreate it (potentially on another executor)